swarm/storage/resource.go

```diff
@@ -0,0 +1,529 @@
+// concatenated in that order, for a total of 16 bytes.
+//
+// Note that the root entry is not required for the resource update scheme to
+// work. A normal chunk of the blocknumber/frequency data can also be created,
+// and pointed to by an actual ENS entry instead.
+//
+// Actual data updates are also made in the form of swarm chunks. The keys
+// of the updates are the hash of a concatenation of properties as follows:
+//
+// sha256(namehash|blocknumber|version)
+//
+// The blocknumber here is the next block period after the current block
+// calculated from the start block and frequency of the resource update.
+// Using our previous example, this means that an update made at block 4285,
+// and even 4284, will have 4326 as the block number.
+//
+// If more than one update is made to the same block number, incremental
+// version numbers are used successively.
+//
+// A lookup agent need only know the identifier name
+//
+// the data itself is prefixed with a signed hash of the data. The sigining key
+// can be used to verify the authenticity of the update, for example by looking
+// up the ownership of the namehash in ENS and comparing to the address derived
+// from it
+//
+// NOTE: the following is yet to be implemented
+// The resource update chunks will be stored in the swarm, but receive special
+// treatment as their keys do not validate as hashes of their data. They are also
+// stored using a separate store, and forwarding/syncing protocols carry per-chunk
+// flags to tell whether the chunk can be validated or not; if not it is to be
+// treated as a resource update chunk.
+//
+// TODO: signature validation
+type ResourceHandler struct {
+	ChunkStore
+	ethapi       *rpc.Client
+	resources    map[string]*resource
+	hashLock     sync.Mutex
+	resourceLock sync.Mutex
+	hasher       SwarmHash
+	privKey      *ecdsa.PrivateKey
+}
+
+// Create or open resource update chunk store
+func NewResourceHandler(privKey *ecdsa.PrivateKey, datadir string, cloudStore CloudStore, ethapi *rpc.Client) (*ResourceHandler, error) {
+	path := filepath.Join(datadir, "resource")
+	dbStore, err := NewDbStore(datadir, nil, singletonSwarmDbCapacity, 0)
+	if err != nil {
+		return nil, err
+	}
+	localStore := &LocalStore{
+		memStore: NewMemStore(dbStore, singletonSwarmDbCapacity),
+		DbStore:  dbStore,
+	}
+	hasher := MakeHashFunc("SHA256")
+	return &ResourceHandler{
+		ChunkStore:   newResourceChunkStore(path, hasher, localStore, cloudStore),
+		ethapi:       ethapi,
+		resources:    make(map[string]*resource),
+		resourceLock: sync.Mutex{},
+		hashLock:     sync.Mutex{},
+		hasher:       hasher(),
+		privKey:      privKey,
+	}, nil
+}
+
+// Creates a standalone resource object
+//
+// Can be passed to SetResource if external root data lookups are used
+func NewResource(name string, startblock uint64, frequency uint64) (*resource, error) {
+	validname, err := idna.ToASCII(name)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
And maybe we could check the frequency here too (although I see that it will be checked in SetResource later, but better catch it early)
Or maybe even better to extract the resource validity checks into a function (the idna safe check of the name is also duplicated), and call this new function from all the places where needed.